`PwRelaxWorkChain`: Revisit work chain logic

mbercx opened the pull request:
Fixes #733
Fixes #705 

Will write full commit message when all changes are agreed upon.

mbercx commented:
@sphuber some initial notes/questions/decisions:

1. Regarding the meta-convergence: 

   * In #705, I forgot to describe one of the most important reasons to more away from using the volume for the meta iterations: although it's rare that the volume condition rarely doesn't catch an issue with Pulay stresses, it _will_ often run another geometry optimization unnecessarily.
   * Regarding the condition on the kpoints mesh not changing: should we be very strict, and expect _the same_ mesh, or is it fine in case the previous run was done with a denser mesh than the output structure would require? Of course kpoints convergence will be achieved in the latter case, but there might be use cases where the user wants the kpoints mesh to approach the desired density as close as possible. For now I've decided to be strict.
 
2. Should the initial relaxation be run by default? Currently it is, and if we don't make it the default I'm wondering how to allow the user to enable it while also being able to provide a protocol for this step.
3. Re the documenation: I was planning to describe the work chain Purpose and Logic in the Topics section. What do you think?

sphuber commented on src/aiida_quantumespresso/workflows/pw/relax.py:
```diff
@@ -192,74 +198,134 @@ def setup(self):
+                }
+            }
+            new_kpts_mesh, _ = create_kpoints_from_distance(**inputs_create_kpoints).get_kpoints_mesh()
+            if tuple(input_kpts_mesh) != tuple(new_kpts_mesh):
```
_EDIT: just saw your point in the PR description. Will leave this here because it contains some more thoughts on the exact criterion_.

Should we maybe here check only that the new grid is not denser? If the new required grid is _less_ dense, then there is no point in rerunning is there? The criterion for `kpoints_distance` that a _minimum_ density is guaranteed in order to guarantee a minimum precision.

One tricky thing I guess might be if you have non isometric meshes, such as `(2, 3, 1)` and `(3, 2, 1)`. Which one is denser? I guess there you would have to compute the actual distance between kpoints in 1/A for both both meshes in each direction and as soon as just one is larger for the new structure, we conclude we have to rerun.

mbercx replied:
I'm not sure how common that case last case is, but it's a fair point. Maybe we can accept when the new mesh is denser along every direction. This would _not_ accept your last use case, but I think that's fine, and simpler than _really_ calculating the density.

sphuber commented on src/aiida_quantumespresso/workflows/pw/relax.py:
```diff
@@ -192,74 +198,134 @@ def setup(self):
-        inputs.pw.structure = self.ctx.current_structure
 
-        # If one of the nested `PwBaseWorkChains` changed the number of bands, apply it here
+        if 'base_init_relax_workchain' in self.ctx:
+            inputs.pw.structure = self.ctx.base_init_relax_workchain.outputs.output_structure
+        else:
+            inputs.pw.structure = self.ctx.current_structure
```
This is not necessary because `self.ctx.current_structure` is correctly set in `inspect_init_relax`

sphuber commented on src/aiida_quantumespresso/workflows/pw/relax.py:
```diff
@@ -269,80 +335,19 @@ def inspect_relax(self):
-
         # Get the latest relax workchain and pass the outputs
-        final_relax_workchain = self.ctx.workchains[-1]
+        final_relax_workchain = self.ctx.base_relax_workchains[-1]
```
You can remove the try-except on line 355. The `self.ctx.workchain_scf` is never set as that used to be done in the final scf.

sphuber commented on src/aiida_quantumespresso/workflows/pw/relax.py:
```diff
@@ -192,74 +198,134 @@ def setup(self):
+            return False
+
+        # Stop if the maximum number of meta iterations has been reached
+        if self.ctx.iteration == self.inputs.max_meta_convergence_iterations.value:
+            self.report('Maximum number of meta convergence iterations reached.')
+            return False
```
I was looking for the logic where this would set an exit code to say the workchain didn't converge, but to my surprise it is not there, and also didn't use to be there. Think we'd better add one, right?

sphuber commented on src/aiida_quantumespresso/workflows/pw/relax.py:
```diff
@@ -31,13 +23,16 @@ def define(cls, spec):
             exclude=('clean_workdir', 'pw.structure', 'pw.parent_folder'),
             namespace_options={'help': 'Inputs for the `PwBaseWorkChain` for the main relax loop.'})
-        spec.expose_inputs(PwBaseWorkChain, namespace='base_final_scf',
+        spec.expose_inputs(PwBaseWorkChain, namespace='base_init_relax',
```
Nitpick: Maybe move this before the `base_relax` namespace since chronologically it will be used before.

sphuber reviewed (commented):
Thanks @mbercx . I think this is looking good. Left some comments.

> Should the initial relaxation be run by default? Currently it is, and if we don't make it the default I'm wondering how to allow the user to enable it while also being able to provide a protocol for this step.

_Is_ it the default? If the `base_init_relax` namespace is not defined at all, it won't be run. Do you maybe mean the default when using `get_builder_from_protocol`? If it works really well for most real use-cases, I think it makes sense to actually use it by default. We can add a note in the documentation how to skip it, which should not be complicated.

> Re the documenation: I was planning to describe the work chain Purpose and Logic in the Topics section. What do you think?

Yeah, makes perfect sense.

mbercx commented on src/aiida_quantumespresso/workflows/protocols/pw/relax.yaml:
```diff
@@ -2,24 +2,27 @@ default_inputs:
                     press_conv_thr: 0.5
-    base_final_scf:
+    base_init_relax:
+        kpoints_distance: 0.3
+        meta_parameters:
+            conv_thr_per_atom: 1.e-6
+            etot_conv_thr_per_atom: 1.e-3
         pw:
             parameters:
                 CONTROL:
-                    calculation: scf
+                    forc_conv_thr: 0.5e-2
+                CELL:
+                    press_conv_thr: 2
```
![side-eye-monkey](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/791ba222-a643-4298-9df7-5c4af02f071b)

mbercx committed:
‼️ `PwRelaxWorkChain`: Revisit work chain logic

Update the `PwRelaxWorkChain` work chain logic to:

1. Remove the final SCF. Now that Quantum ESPRESSO runs a final SCF at the end of every
   `vc-relax`, there is no need to do this in the workflow.
2. Adapt the meta-convergence condition. Instead of running an extra geometry
   optimization based on the volume difference (which always requires two runs), we:

   * Make sure there are no Pulay stresses by checking the stress result from the
     QE-run final SCF, which resets the basis sets based on the unit cell after the
     geometry optimization.
   *  Make sure the k-point mesh is still sufficiently dense to adhere to the
     `kpoints_distancy` input.

3. Add an initial relaxation with looser precision settings:

   * This makes it less costly to obtain a reasonable first ground state geometry,
     improving the work chain efficiency
   * The less strict parameters makes it easier for the SCF to converge for geometries
     far removed from the ground state, improving robustness.

Minotakm commented:
Probably I have done some changes while I was running it. I will try to find some time next week to check.  Thanks for letting me know.

mbercx commented:
Thanks @Minotakm! As mentioned, I've merged the PR for now to do an alpha release, but comments or changes are very welcome! I'll be working on some other parts of the code and gathering feedback on any breaking changes, so we can get the best API possible before doing a stable major release.

mbercx commented:
@bastonero I remember you also mentioned that you liked the previous work chain version, and wanted to keep it? I'm not keen on maintaining two versions of the `PwRelaxWorkChain`, but I would consider moving it to a separate package for "unsupported" work chains. Thoughts welcome!

bastonero commented:
@mbercx thanks for pinging me! Honestly, i don't think it is so much effort to keeping both. Moreover, they do two different things, so I think with a correct naming they could be both of interests.
